### DISCUSSION TITLE
Switch network without a wallet connected?

### DISCUSSION THREAD

**Discussion by noahlitvin** [wevm/wagmi#461] [Q&A] [answered]
Hi! Thanks for building this.

I'm interested in allowing the user to change their network, such that the `useContractRead` hooks will pull from the appropriate provider, without requiring a wallet to be connected. This sort of functionality is available in the Uniswap UI, for example.

What's the best approach here? It seems like `switchNetwork` relies on a connector. Maybe we could update `useProvider` to return `{provider, setProvider}`? The network switcher would render options from `client.chains` and call `setProvider` with the selected chain's id as a param?

Happy to get a PR started with some guidance. Or is there a solution without an update to the library?

**Comment by tmm** [comment, marked as answer]
All query hooks (e.g. `useContractRead`) have a `chainId` parameter. You can pass whatever chain id you want and the query will use that when fetching data.

Based on your requirements, I would keep track of the chain id in some global store (e.g. React Context, Redux, Zustand) and pass that value to hooks. You could create a `<NetworkSwitcher />` component that uses `useNetwork`s `switchNetwork` when connected and reads/writes to global store when disconnected.

**Reply by noahlitvin** [reply]
Thanks! I can move in this direction, but it'd be great to avoid adding this on every useContractRead call. Out of curiosity, is having some sort of `setProvider` functionality on the roadmap? I see how this might be a non-trivial change.

For a little more context, I'm looking at setting this up with Rainbow Kit. In [this example](https://www.rainbowkit.com/docs/custom-connect-button), `chain` is `activeChain` [here](https://github.com/rainbow-me/rainbowkit/blob/main/packages/rainbowkit/src/components/ConnectButton/ConnectButtonRenderer.tsx#L92). I guess I'd expect that value to always return the chain that a `useContractRead` call would query. I'd look at `activeConnector` on `useConnect` to check for a connected wallet.

**Reply by noahlitvin** [reply]
cc @peduarte @markdalgleish - Do you have a suggested approach for network switching without a connected wallet?

**Reply by tmm** [reply]
Nothing on the roadmap right now. Would need to answer: What chains are used when there is no connection? What happens when a wallet is connected/disconnected? There are probably additional implications as well.

Something I just thought of that you could try is using `client`s `setState` function to set the chain id (e.g. `client.setState(x => ({ ...x, data: { ...x.data, chainId }))`). No guarantee this will work in the future though.

**Reply by noahlitvin** [reply]
Gotcha. I guess I'd expect it to always have a value, defaulting to mainnet if unset, per [ethers](https://docs.ethers.io/v5/api/providers/#providers-getDefaultProvider). I think this is 'lower level' than wallet connection, but if the connector switches networks, this would stay synced. I assume my use case is going to become the default expected UI, so happy to contribute to a solution here.

And I appreciate the tip! I'll keep you posted. Thanks again.

**Reply by ankerx** [reply]
Did you find a solution? @noahlitvin
